timemachine/md/barker.py

```diff
@@ -0,0 +1,61 @@
+import numpy as np
+
+
+class BarkerProposal:
+    def __init__(self, grad_log_q, proposal_sig=0.001):
+        """Robust gradient-informed proposal distribution.
+
+        Supports:
+        * sampling from proposal y ~ p(. | x)
+        * evaluation of proposal density p(y | x)
+
+        Compared to Langevin proposals, more robust to poor initialization and to poor choice of step size.
+
+        References
+        ----------
+        [Livingstone, Zanella, 2020]
+            The Barker proposal: combining robustness and efficiency in gradient-based MCMC
+            https://arxiv.org/abs/1908.11812
+        """
+        self.grad_log_q = grad_log_q
+        assert proposal_sig > 0
+        self.proposal_sig = proposal_sig
+
+    def _sample(self, x, gaussian_rvs, uniform_rvs):
```

Comment:
To clarify, `uniform_rvs` is expected to be a normal distribution? 

In `sample()` the gaussian_rvs and uniform_rvs both come from `randn` (https://numpy.org/doc/stable/reference/random/generated/numpy.random.randn.html) which are both normal.

Reply:
>To clarify, uniform_rvs is expected to be a normal distribution?

No -- expected to be uniform

>In sample() the gaussian_rvs and uniform_rvs both come from randn (https://numpy.org/doc/stable/reference/random/generated/numpy.random.randn.html) which are both normal.

No -- uniform_rvs come from `rand`, which is uniform in [0,1)

https://github.com/proteneer/timemachine/blob/743706d4972fe3adddc1c4a62b1950986b97589b/timemachine/md/barker.py#L42-L47

Reply:
🤦 Just can't read

Reply:
(Reason for splitting it up is so that `_sample` can be a deterministic function. Isolating the RNG calls seems cleaner in my opinion (more easily swap global numpy RNG vs. jax RNG if desired, the deterministc function may be more easily tested...), and could be useful for batching, which could look like the following if `grad_log_q` is amenable to jax transformations)

```python
    def batch_sample(self, xs):
        """[y_i ~ p(. | x_i) for x_i in xs], using vmap"""
        gauss = np.random.randn(*xs.shape)
        unif = np.random.rand(*xs.shape)

        return vmap(self._sample, (0, 0, 0))(xs, gauss, unif)
```

(batching was reverted in https://github.com/proteneer/timemachine/pull/978/commits/e2773b00dea061dc9305f5e1adf2001ee3efa9ff to keep scope smaller, and to avoid making `barker.py` depend on jax initially)